src/main/java/dev/morling/onebrc/CalculateAverage_merykitty.java

```diff
@@ -0,0 +1,368 @@
+        }
+
+        static int rehash(int x) {
+            x = ((x >>> 16) ^ x) * 0x45d9f3b;
+            x = ((x >>> 16) ^ x) * 0x45d9f3b;
+            x = (x >>> 16) ^ x;
+            return x;
+        }
+
+        private boolean keyEqualScalar(PoorManMapNode node, long offset, long size, int hash) {
+            if (node.hash != hash || node.size != size) {
+                return false;
+            }
+
+            // Be simple
+            for (int i = 0; i < size; i++) {
+                int c1 = node.data[BYTE_SPECIES.vectorByteSize() + i];
+                int c2 = data.get(ValueLayout.JAVA_BYTE, offset + i);
+                if (c1 != c2) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    private static long parseDataPoint(Aggregator aggr, MemorySegment data, long offset) {
+        long word = data.get(JAVA_LONG_LT, offset);
+        // This can be 12, 20, 28
+        int decimalSepPos = Long.numberOfTrailingZeros(~word & 0x10101000);
+        int shift = 28 - decimalSepPos;
+        // signed is -1 if negative, 0 otherwise
+        long signed = (~word << 59) >> 63;
+        long designMask = ~(signed & 0xFF);
+        long digits = ((word & designMask) << shift) & 0x0F000F0F00L;
+
+        // Now digits is in the form 0xUU00TTHH00
+        // 0xUU00TTHH00 * (100 * 0x1000000 + 10 * 0x10000 + 1) =
+        // 0x000000UU00TTHH00 +
+        // 0x00UU00TTHH000000 * 10 +
+        // 0xUU00TTHH00000000 * 100
+        // Now TT * 100 has 2 trailing zeroes and HH * 100 + TT * 10 + UU < 0x400
+        // This results in our value lies in the bit 32 to 41 of this product
+        // That was close :)
+        long value = ((digits * 0x640a0001) >>> 32) & 0x3FF;
+        int point = (int) ((value ^ signed) - signed);
+        aggr.min = Math.min(point, aggr.min);
+        aggr.max = Math.max(point, aggr.max);
+        aggr.sum += point;
+        aggr.count++;
+        return offset + (decimalSepPos >>> 3) + 3;
+    }
+
+    private static long parseDataPointTail(Aggregator aggr, MemorySegment data, long offset) {
+        int point = 0;
+        boolean negative = false;
+        if (data.get(ValueLayout.JAVA_BYTE, offset) == '-') {
+            negative = true;
+            offset++;
+        }
+        for (;; offset++) {
+            int c = data.get(ValueLayout.JAVA_BYTE, offset);
+            if (c == '.') {
+                c = data.get(ValueLayout.JAVA_BYTE, offset + 1);
+                point = point * 10 + (c - '0');
+                offset += 3;
+                break;
+            }
+
+            point = point * 10 + (c - '0');
+        }
+        point = negative ? -point : point;
+        aggr.min = Math.min(point, aggr.min);
+        aggr.max = Math.max(point, aggr.max);
+        aggr.sum += point;
+        aggr.count++;
+        return offset;
+    }
+
+    private static long iterate(PoorManMap aggrMap, MemorySegment data, long offset) {
```

Comment:
Could you provide a description of how that iteration loop works? I roughly get it, but it's quite dense code and having a high-level description would be great.

Reply:
Done